### ISSUE TITLE
Error Loading module hook 'hook-_tkinter.py'

### ISSUE THREAD

**Issue by Biboba** [pyinstaller/pyinstaller#7669]
## Description of the issue

I can't build my application anymore since I upgraded from Windows 10 to Windows 11. Python dependencies are freezed with pip so the only thing that changed is my OS.

### Context information (for bug reports)

* Output of `pyinstaller --version`: ```5.11.0```
* Version of Python: 3.9
* Platform: Windows 11 in French 
* How you installed Python: conda
* Did you also try this on another platform? No

* try the latest development version, using the following command:

```shell
pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip
```

* follow *all* the instructions in our "If Things Go Wrong" Guide
  (https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong) and

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [X] start with clean installation
  * [X] use the latest development version
  * [x] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [X] Package your program in **--onedir mode**
  * [X] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [x] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### A minimal example program which shows the error

- Install 'arcgis' package with pip: 
```
pip install arcgis
```

- In main.py
```
import arcgis
print(arcgis.__version)
```

- Run pyinstaller:
```
pyinstaller main.py --log-level DEBUG
```

### Stacktrace / full error message

```
76692 INFO: Loading module hook 'hook-tinycss2.py' from 'C:\\Users\\foo\\AppData\\Local\\ESRI\\conda\\envs\\py\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
76692 DEBUG: Collecting data files for tinycss2
76696 DEBUG: collect_data_files - Found files: []
76844 WARNING: Hidden import "importlib_resources.trees" not found!
76849 INFO: Loading module hook 'hook-matplotlib.backends.py' from 'C:\\Users\\foo\\AppData\\Local\\ESRI\\conda\\envs\\py\\lib\\site-packages\\PyInstaller\\hooks'...
76850 INFO: Matplotlib backend selection method: automatic discovery of used backends
77180 INFO: Trying determine the default backend as first importable candidate from the list: ['Qt5Agg', 'Gtk3Agg', 'TkAgg', 'WxAgg']
78481 INFO: None of the backend candidates could be imported; falling back to headless Agg!
78481 INFO: Selected matplotlib backends: ['Agg']
78876 INFO: Loading module hook 'hook-PIL.SpiderImagePlugin.py' from 'C:\\Users\\foo\\AppData\\Local\\ESRI\\conda\\envs\\py\\lib\\site-packages\\PyInstaller\\hooks'...
78882 DEBUG: Suppressing import of 'tkinter' from module 'PIL.ImageTk' due to excluded import 'tkinter' specified in a hook for 'PIL.ImageTk' (or its parent package(s)).
81260 INFO: Loading module hook 'hook-setuptools.msvc.py' from 'C:\\Users\\foo\\AppData\\Local\\ESRI\\conda\\envs\\py\\lib\\site-packages\\PyInstaller\\hooks'...
81262 DEBUG: Suppressing import of 'numpy' from module 'setuptools.msvc' due to excluded import 'numpy' specified in a hook for 'setuptools.msvc' (or its parent package(s)).
82547 INFO: Loading module hook 'hook-_tkinter.py' from 'C:\\Users\\foo\\AppData\\Local\\ESRI\\conda\\envs\\py\\lib\\site-packages\\PyInstaller\\hooks'...
Traceback (most recent call last):
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\Scripts\pyinstaller.exe\__main__.py", line 7, in <module>
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\__main__.py", line 194, in _console_script_run
    run()
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\__main__.py", line 180, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\__main__.py", line 61, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\building\build_main.py", line 1006, in main
    build(specfile, distpath, workpath, clean_build)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\building\build_main.py", line 928, in build
    exec(code, spec_namespace)
  File "C:\temp\pyinstaller-bug\main.spec", line 7, in <module>
    a = Analysis(
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\building\build_main.py", line 428, in __init__
    self.__postinit__()
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\building\datastruct.py", line 184, in __postinit__
    self.assemble()
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\building\build_main.py", line 587, in assemble
    self.graph.process_post_graph_hooks(self)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\depend\analysis.py", line 329, in process_post_graph_hooks
    module_hook.post_graph(analysis)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\depend\imphook.py", line 447, in post_graph
    self._process_hook_func(analysis)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\depend\imphook.py", line 469, in _process_hook_func
    self._hook_module.hook(hook_api)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\hooks\hook-_tkinter.py", line 27, in hook
    hook_api.add_datas(collect_tcl_tk_files(hook_api.__file__))
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\utils\hooks\tcl_tk.py", line 243, in collect_tcl_tk_files
    tcl_root, tk_root = _find_tcl_tk(tkinter_ext_file)
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\site-packages\PyInstaller\utils\hooks\tcl_tk.py", line 205, in _find_tcl_tk
    tk_dir = os.path.join(os.path.dirname(tcl_dir), f"tk{tk_version}")
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\ntpath.py", line 223, in dirname
    return split(p)[0]
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\ntpath.py", line 185, in split
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Thanks for your help !

Please also see <https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs>
for more about what would use to solve the issue.

**Comment by rokm** [comment]
What happens if you run the following code snippet in your python environment?

```
import tkinter
from _tkinter import TCL_VERSION, TK_VERSION

tcl = tkinter.Tcl()
tcl_dir = tcl.eval("info library")

print(tcl_dir)
```

**Comment by Biboba** [comment]
Thanks for your reply.
It's failing on:
```
import tkinter
```
error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\foo\AppData\Local\ESRI\conda\envs\py\lib\tkinter\__init__.py", line 37, in <module>
    import _tkinter # If this fails your Python may not be configured for Tk
ImportError: DLL load failed while importing _tkinter: Le module spécifié est introuvable.
```

**Comment by rokm** [comment]
So `tkinter` cannot be imported because either `_tkinter` is unavailable (which is unlikely, because the hook for it was fired) or because one of its dependencies is unavailable.

`_tkinter.pyd` should be in `C:\Users\foo\AppData\Local\ESRI\conda\envs\py\DLLs\_tkinter.pyd`. The same DLLs directory should also contain `tcl86t.dll` and `tk86t.dll`. Is that the case?

While that `None`-related error could be handled gracefully in the hook utility function, you would hit an error at run-time due to checks in the `_tkinter` run-time hook (which is activated, because some part of the collected code imports `tkinter`). So you should figure out what the problem with `tkinter` in that environment is and fix it. Alternatively, if you are 100% you don't need `tkinter` (i.e., it is included because some part of collected code is conditionally importing it, but you don't need that part), you can explicitly exclude it by adding `--exclude tkinter --exclude _tkinter` to your PyInstaller command (or add `tkinter` and `_tkinter` to `excludes` argument of `Analysis` in the .spec file).

**Comment by Biboba** [comment]
- I confirm `_tkinter.pyd`,`tcl86t.dll` and `tk86t.dll` are located in `C:\Users\foo\AppData\Local\ESRI\conda\envs\py\DLLs` folder

- The thing that I find strange is that I am facing the issue since I change OS from Windows 10 to Windows 11. I did not touch the version of the [arcgis](https://pypi.org/project/arcgis/) library I am using. And a grep for 'tkinter' or '_tkinter' in the `arcgis` package located in `C:\Users\foo\AppData\Local\ESRI\conda\envs\py\Lib\site-packages\arcgis` does show any result.

- I confirm that adding `--exclude tkinter --exclude _tkinter` solve my issue

Thanks !

**Comment by rokm** [comment]
> * The thing that I find strange is that I am facing the issue this I change OS from Windows 10 to Windows 11. I did not touch the version of the [arcgis](https://pypi.org/project/arcgis/) library I am using. And a grep for 'tkinter' or '_tkinter' in the `arcgis` package located in `C:\Users\foo\AppData\Local\ESRI\conda\envs\py\Lib\site-packages\arcgis` does show any result

This doesn't have to do anything with `arcgis` (unless `arcgis` depends on `tkinter`, but even then it would likely be unrelated); it looks like an issue with missing (or incompatible) DLL that should be part of your anaconda-installed python. Could be that a DLL is missing due to Windows 10 to 11 upgrade, but I thought anaconda environment provided its own copies of base DLLs exactly for cases like that.

If you create a new anaconda environment with the same python version and activate it, are you able to import `tkinter` there? What if you create environment with the latest python version?

**Comment by Biboba** [comment]
It seems like it depends on Python version indeed:
- Anaconda environment created with Python 3.10 and 3.11: `import tkinter` works
- Anaconda environment created with Python 3.9: `import tkinter` fails with the DLL error above

**Comment by bwoodsend** [comment]
Probably fixed by #7671? Does `pip install -U pyinstaller` make this issue go away?

**Comment by rokm** [comment]
> Probably fixed by #7671? Does `pip install -U pyinstaller` make this issue go away?

Nah, this is not even PyInstaller-specific. It looks like `tkinter` on some Anaconda python versions (3.9) is not compatible with Windows 11.

**Comment by Biboba** [comment]
Yes, indeed, turns out not to be specific to Pyinstaller.
Many thanks @rokm and @bwoodsend for the great support and follow-up 🙏 !
